{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Various script used mainly in the preprocesssing of the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from os import listdir\n",
    "import os\n",
    "from os.path import isfile, join\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "folder_path = \"dataset/topo/embeddings/dnabert\"\n",
    "\n",
    "\n",
    "# list of files in the folder without the extension\n",
    "IDs = [Path(f).stem for f in listdir(folder_path) if isfile(join(folder_path, f))]\n",
    "\n",
    "\n",
    "embeddings_matrix = []\n",
    "\n",
    "\n",
    "for id in IDs:\n",
    "\n",
    "    # load the file as a numpy array\n",
    "    raw_embedding = None\n",
    "    try:\n",
    "        raw_embedding = np.load(folder_path + \"/\" + id + \".npy\")\n",
    "        print(raw_embedding.shape)\n",
    "    except:\n",
    "        raw_embedding = np.load(folder_path + \"/\" + id + \".npy\", allow_pickle=True)\n",
    "        print(f\"Error while loading the embedding of sequence {id}\")\n",
    "        # os.remove(folder_path + \"/\" + id + \".npy\")\n",
    "        print(raw_embedding.shape)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "\n",
    "\n",
    "FASTA_FILE_PATH = \"/storagenfs/m.tolloso/BioEmbedding/dataset/satb2/satb2_tr.txt\"\n",
    "lens = []\n",
    "\n",
    "for count, seqrecord in enumerate(SeqIO.parse(FASTA_FILE_PATH, \"fasta\")):\n",
    "\n",
    "    lens.append(len(str(seqrecord.seq)))\n",
    "\n",
    "print(count + 1)\n",
    "\n",
    "print(np.mean(lens))\n",
    "print(np.max(lens))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export the geneIDs given a xml uniprot file\n",
    "from Bio import SeqIO\n",
    "\n",
    "records = list(SeqIO.parse(\"/storagenfs/m.tolloso/BioEmbedding/dataset/hemoglobin/hemoglobin.xml\", \"uniprot-xml\"))\n",
    "geneIDs = []\n",
    "\n",
    "with_gid = 0\n",
    "\n",
    "xml_names = []\n",
    "for record in records:\n",
    "    for i in record.dbxrefs:\n",
    "        if i.startswith(\"GeneID:\"):\n",
    "            geneID = i.split(\":\")[1]\n",
    "            geneIDs.append(geneID)\n",
    "            with_gid += 1\n",
    "            break\n",
    "    \n",
    "\n",
    "\n",
    "print(with_gid)\n",
    "\n",
    "# with open(\"emoglobina_ids.txt\", 'w') as f:\n",
    "#     for i in geneIDs:\n",
    "#         f.write(i + \"\\n\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given a genbank(full) file, for each proteome select the superset proteins (CDS) and add them to a fasta file\n",
    "\n",
    "from Bio import SeqIO\n",
    "import Bio\n",
    "\n",
    "no_ids = 0\n",
    "\n",
    "def to_simple_location(f_list):\n",
    "    for feature in f_list: \n",
    "        if type(feature.location) == Bio.SeqFeature.CompoundLocation:\n",
    "            feature.location = (int(feature.location.parts[0].start) , int(feature.location.parts[1].end))\n",
    "        else:\n",
    "            feature.location = (int(feature.location.start), int(feature.location.end))\n",
    "\n",
    "def remove_subsets(f_list):\n",
    "    for i in range(len(f_list)):\n",
    "        for j in range(i+1, len(f_list)):\n",
    "            if f_list[i].location[0] <= f_list[j].location[0] and f_list[i].location[1] >= f_list[j].location[1]:\n",
    "                # print(f'removed protein {f_list[j].qualifiers[\"protein_id\"]}')\n",
    "                f_list.pop(j)\n",
    "                break\n",
    "            elif f_list[j].location[0] <= f_list[i].location[0] and f_list[j].location[1] >= f_list[i].location[1]:\n",
    "                # print(f'removed protein {f_list[i].qualifiers[\"protein_id\"]}')\n",
    "                f_list.pop(i)\n",
    "                break\n",
    "            else:\n",
    "                continue\n",
    "        break\n",
    "    return f_list\n",
    "\n",
    "\n",
    "gb_records = list(SeqIO.parse('dataset/covid19/covid19.gb','genbank'))\n",
    "final_proteins = []\n",
    "\n",
    "for gb_record in gb_records:\n",
    "    CDS_features = []\n",
    "\n",
    "    for feature in gb_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            CDS_features.append(feature) \n",
    "\n",
    "    to_simple_location(CDS_features)\n",
    "\n",
    "    # print(len(CDS_features))\n",
    "    remove_subsets(CDS_features)\n",
    "    # print(len(CDS_features))\n",
    "\n",
    "    for feature in CDS_features:\n",
    "        # print(feature.location, feature.qualifiers['protein_id'], feature.qualifiers['translation'])\n",
    "\n",
    "        try:\n",
    "            final_proteins.append((feature.qualifiers['protein_id'], feature.qualifiers['translation']))\n",
    "        except:\n",
    "            no_ids +=1\n",
    "\n",
    "\n",
    "print(len(final_proteins))\n",
    "print(no_ids)  \n",
    " \n",
    "\n",
    "# output_file = f\"dataset/covid19/covid19_proteins.fasta\"\n",
    "\n",
    "# with open(output_file, 'w') as fasta_file:\n",
    "#     for entry in final_proteins:\n",
    "#         description, sequence = entry\n",
    "#         fasta_file.write(f'>{description[0]}\\n{sequence[0]}\\n')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "\n",
    "gb_records = list(SeqIO.parse('/storagenfs/m.tolloso/BioEmbedding/dataset/meningite/meningite_tr.fasta','fasta'))\n",
    "\n",
    "lens = []\n",
    "\n",
    "for gb_record in gb_records:\n",
    "    lens.append(len(gb_record.seq))\n",
    "\n",
    "print(np.min(lens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import numpy as np \n",
    "\n",
    "path = '/storagenfs/m.tolloso/BioEmbedding/dataset/covid19/embeddings/alphafold'\n",
    "\n",
    "# for each file in the folder 'path'\n",
    "\n",
    "ids = os.listdir(path)\n",
    "\n",
    "\n",
    "tensor1 = np.load(path + '/' + ids[0])\n",
    "tensor2 = np.load(path + '/' + ids[3])\n",
    "\n",
    "print(tensor1.shape)\n",
    "print(tensor2.shape)\n",
    "\n",
    "# concat the tensors over the second axis\n",
    "\n",
    "concat = np.concatenate((tensor1, tensor2), axis=1)\n",
    "\n",
    "print(concat.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "import numpy as np \n",
    "\n",
    "records = list(SeqIO.parse('dataset/batterio/batterio_tr.fasta','fasta'))\n",
    "\n",
    "geneIDS = []\n",
    "\n",
    "for record in records:\n",
    "    id = record.id\n",
    "    geneID = record.description.split(\"[GeneID=\")[1].split(']')[0]\n",
    "    # geneIDS.append(geneID)\n",
    "\n",
    "    os.rename(f'dataset/batterio/embeddings/dnabert/{id}.npy', f'dataset/batterio/embeddings/dnabert/{geneID}.npy')\n",
    "    # print(id, geneID)\n",
    "\n",
    "# print(len(geneIDS))\n",
    "# print(len(set(geneIDS)))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "def extract_gene_id(fasta_header):\n",
    "    match = re.search(r'\\[GeneID=(\\d+)\\]', fasta_header)\n",
    "    return match.group(1) if match else None\n",
    "\n",
    "\n",
    "def process_fasta_file(input_file_path, output_file_path):\n",
    "    with open(input_file_path, 'r') as input_file, open(output_file_path, 'w') as output_file:\n",
    "        for line in input_file:\n",
    "            if line.startswith('>'):\n",
    "\n",
    "                gene_id = extract_gene_id(line)\n",
    "                \n",
    "                # Modify the header to include GeneID\n",
    "                modified_header = line.replace('>', f'>{gene_id} ')\n",
    "                output_file.write(modified_header)\n",
    "                \n",
    "                # Update current header for the new sequence\n",
    "                current_header = line.strip()\n",
    "            else:\n",
    "                # Continue writing sequence lines\n",
    "                output_file.write(line)\n",
    "\n",
    "\n",
    "\n",
    "# Replace 'input.fasta' and 'output.fasta' with the actual file paths\n",
    "input_file_path = '/storagenfs/m.tolloso/BioEmbedding/dataset/batterio/batterio_tr.fasta'\n",
    "output_file_path = '/storagenfs/m.tolloso/BioEmbedding/dataset/batterio/batterio_tr_new.fasta'\n",
    "process_fasta_file(input_file_path, output_file_path)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "\n",
    "records = list(SeqIO.parse('dataset/mouse/mouse_tr.fasta','fasta'))\n",
    "files = os.listdir('dataset/mouse/embeddings/dnabert')\n",
    "\n",
    "ids = [record.id for record in records]\n",
    "embs = [file.split('.')[0] for file in files]\n",
    "\n",
    "\n",
    "for id in ids:\n",
    "    if id not in embs:\n",
    "        print(id)  \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aggegate the satb fasta files\n",
    "from Bio import SeqIO\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "files = ['Satb1_Orthologues_protein_sequences.txt', 'Satb1_Satb2_protein_sequences.txt', 'Satb2_Orthologues_protein_sequences.txt']\n",
    "files_tr = ['Satb1_Orthologues_tr_sequences.txt', 'Satb1_Satb2_tr_sequences.txt', 'Satb2_Orthologues_tr_sequences.txt']\n",
    "\n",
    "transcript_dict = {}\n",
    "\n",
    "for file_tr in files_tr:\n",
    "    records_fasta = list(SeqIO.parse(\"/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/bacterium_proteins.faa\", \"fasta\"))\n",
    "\n",
    "    for record_tr in records_fasta:\n",
    "        transcript_dict[record_tr.id] = record_tr.description.split('[GeneID=')[1][:-1]\n",
    "\n",
    "\n",
    "\n",
    "embedder = 'alphafold'\n",
    "\n",
    "\n",
    "\n",
    "files = os.listdir(f'/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/{embedder}')\n",
    "for file in files :\n",
    "\n",
    "    # print(f'/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/{embedder}/{protein_name}.npy')\n",
    "    # print(f'/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/{embedder}/{transcript_dict[protein_name]}.npy')\n",
    "\n",
    "\n",
    "    os.rename(f'/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/{embedder}/{file}', \n",
    "                f'/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/{embedder}/{file}'[:-8])\n",
    "\n",
    "          \n",
    "\n",
    "transcript_dict\n",
    "\n",
    "# with open('dataset/satb2/satb2_list.txt', 'w') as of:\n",
    "\n",
    "#     for file in files:\n",
    "#         records = list(SeqIO.parse(f'dataset/satb2/{file}', 'fasta'))\n",
    "        \n",
    "#         for record in records:\n",
    "            \n",
    "#             if len(record.seq) == 0:\n",
    "#                 of.write(f'>gene:{record.id}')\n",
    "#                 seq = transcript_dict[record.id]\n",
    "#             else:\n",
    "#                 of.write(f' protein:{record.id}\\n')\n",
    "#                 of.write(f'{seq}\\n')\n",
    "                 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "file_name = f'dataset/satb2/satb2_tr.txt'\n",
    "\n",
    "import FastaValidator\n",
    "\n",
    "return_code = FastaValidator.fasta_validator(file_name)\n",
    "print(return_code)\n",
    "\n",
    "\n",
    "\n",
    "records_tr = list(SeqIO.parse(f'dataset/satb2/satb2_tr.txt', 'fasta'))\n",
    "\n",
    "\n",
    "ids = []\n",
    "\n",
    "for record in records_tr:\n",
    "    print(record.id)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export the geneIDs given a xml uniprot file\n",
    "from Bio import SeqIO\n",
    "from autoembedding.utils import get_gene_id\n",
    "import os\n",
    "\n",
    "# records_xml = list(SeqIO.parse(\"/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/Neisseria_meningitidis.xml\", \"uniprot-xml\"))\n",
    "records_fasta = list(SeqIO.parse(\"/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/bacterium_tr_new.fasta\", \"fasta\"))\n",
    "\n",
    "\n",
    "\n",
    "complete = 0\n",
    "\n",
    "# for record in records_fasta:\n",
    "#     fasta_IDs.append(record.id)\n",
    "\n",
    "for record in records_fasta:\n",
    "\n",
    "\n",
    "    os.path.exists(f\"/storagenfs/m.tolloso/BioEmbedding/dataset/bacterium/embeddings/alphafold/{record.id}.npy\")\n",
    "    complete += 1\n",
    "\n",
    "\n",
    "print(complete)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final = list(set(xml_IDs).intersection(set(fasta_IDs)))\n",
    "len(final)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "with open('dataset/bacterium/bacterium_tr_new.fasta', 'w') as f:\n",
    "\n",
    "    for record in records_fasta:\n",
    "        if record.id in final:\n",
    "            # write the record to the output file\n",
    "            SeqIO.write(record, f, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from itertools import combinations\n",
    "\n",
    "\n",
    "command = './clustalo-1.2.4-Ubuntu-x86_64 --infile tmp/test.fasta --threads 40 --MAC-RAM 8000 -v --verbose --guidetree-out tmp/gt.dnd --distmat-out tmp/dm.matrix --outfmt clustal --resno --outfile tmp/of.clustal_num --output-order tree-order --seqtype dna'\n",
    "command_nodm = './clustalo-1.2.4-Ubuntu-x86_64 --infile tmp/test.fasta --threads 40 -v --verbose --auto --outfile tmp/of.clustal_num  --seqtype dna'\n",
    "\n",
    "\n",
    "mengite_tr = list(SeqIO.parse(\"/storagenfs/m.tolloso/BioEmbedding/dataset/meningite/meningite_tr.fasta\", \"fasta\"))\n",
    "\n",
    "\n",
    "\n",
    "for record1, record2 in combinations(mengite_tr, 2):\n",
    "    with open(\"test.fasta\", \"w\") as f:\n",
    "        SeqIO.write([record1, record2], f, \"fasta\")\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pickle\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "res = pickle.load(open(\"tmp/res_covid19.pkl\", \"rb\"))\n",
    "Z = res['embeddings_linkage_matrix']\n",
    "leaf_names = res['embeddings_IDs']\n",
    "\n",
    "\n",
    "def get_newick(node, parent_dist, leaf_names, newick='') -> str:\n",
    "    \"\"\"\n",
    "    Convert sciply.cluster.hierarchy.to_tree()-output to Newick format.\n",
    "\n",
    "    :param node: output of sciply.cluster.hierarchy.to_tree()\n",
    "    :param parent_dist: output of sciply.cluster.hierarchy.to_tree().dist\n",
    "    :param leaf_names: list of leaf names\n",
    "    :param newick: leave empty, this variable is used in recursion.\n",
    "    :returns: tree in Newick format\n",
    "    \"\"\"\n",
    "    if node.is_leaf():\n",
    "        return \"%s:%.2f%s\" % (leaf_names[node.id], parent_dist - node.dist, newick)\n",
    "    else:\n",
    "        if len(newick) > 0:\n",
    "            newick = \"):%.2f%s\" % (parent_dist - node.dist, newick)\n",
    "        else:\n",
    "            newick = \");\"\n",
    "        newick = get_newick(node.get_left(), node.dist, leaf_names, newick=newick)\n",
    "        newick = get_newick(node.get_right(), node.dist, leaf_names, newick=\",%s\" % (newick))\n",
    "        newick = \"(%s\" % (newick)\n",
    "        return newick\n",
    "\n",
    "tree = hierarchy.to_tree(Z, False)\n",
    "newick = get_newick(tree, tree.dist, leaf_names)\n",
    "with open(\"results_trees/covid19_esm.nwk\", \"w\") as f:\n",
    "    f.write(newick)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check the proteins of the meningits (for length variance etc.)\n",
    "from Bio import SeqIO\n",
    "\n",
    "gb_records = list(SeqIO.parse('dataset/meningite/meningite.gb','genbank'))\n",
    "proteins = {}\n",
    "\n",
    "for gb_record in gb_records[:2]:\n",
    "\n",
    "    for feature in gb_record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            try:\n",
    "                protein_sequence = feature.qualifiers['translation'][0]\n",
    "                protein_id = feature.qualifiers['product'][0]\n",
    "            except:\n",
    "                continue\n",
    "            try:\n",
    "                proteins[protein_id] += 1\n",
    "            except:\n",
    "                proteins[protein_id] = 1\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "print(len(proteins))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for protein in proteins.keys():\n",
    "\n",
    "    print(protein, (proteins[protein]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the phylogenetic distance matrix, build a UPGMA phylogenetic tree, and export it as a newik tree\n",
    " \n",
    "from autoembedding import utils\n",
    "from scipy.spatial.distance import squareform\n",
    "from scipy.cluster.hierarchy import linkage\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "case_study = \"satb2\"\n",
    "\n",
    "method = \"average\"\n",
    "metric = \"euclidean\"\n",
    "\n",
    "gtrue_IDs, gt_distances = utils.read_distance_matrix(case_study)\n",
    "gt_distances = squareform(gt_distances)\n",
    "gtrue_linkage_matrix = linkage(gt_distances, method=method, metric=metric)\n",
    "\n",
    "tree = hierarchy.to_tree(gtrue_linkage_matrix, False)\n",
    "\n",
    "newick = utils.get_newick(tree, tree.dist, gtrue_IDs)\n",
    "\n",
    "with open(f\"results_trees/{case_study}_upgma.nwk\", \"w\") as f:\n",
    "    f.write(newick)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bioemb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
