\documentclass[12pt, letterpaper, twocolumn]{article}
\usepackage[utf8]{inputenc}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{placeins}


\title{Title}
\author{Author}
\date{Date}



\begin{document}
\maketitle

\section{Introduction}

\section{Related Work}

\section{Methods}

In this section we describe the pipeline used to analyze the embeddings. As shown in Table \ref{table:data}, the input length is different between the models as well as the output produced. We want to address the following problems: 1) compare different methods to join togheter the amminoacid-specific contextual representations in order to have a representation for the whole chunk; 2) compare different methods to join togheter the representations of the chunks in order to have a representation for the whole protein; 3) find out if these representations reflect known properties of the proteins.

\subsection{Combining the contextual representations}
We tried four methods to join togheter the amminoacid embeddings in order to produce a fixed size embedding for the chunk: average, maximum, sum and principal component analysis (PCA). Note that even if these operators are commutative, the overall process do takes into account the order of the amminoacids precisely because the embeddings are contextual.

The same operator used to combine the amminoacid embeddings is also used to combine the embeddings of the chunks of the sequence.

\subsection{Comparison with known informations}
Given a set of embeddings of sequences we want to analyze their distribution in the embedding space comparing it with both the distance matrix produced during the multiple sequence alignment with Clustal Omega \cite{sievers2011fast} and protein annotations as gene ontology, UniProt keywords and taxonimy \cite{uniprot23}.

\subsubsection{Allignment distance}
In order to compare the distances between the sequences in the embedding space with the alignment distance we performed an agglomerative clustering on both the matrices, the resulting tree is then cut at each levels: flat partitions of all possibles number of clusters are produced. We perform a comparison of the partitions with the same number of clusters using the adjusted rand score \cite{hubert1985comparing}. The mean of these score, starting from two clusters up to $ \#elements - 1 $ clusters is called mean adjusted rand score (MARS). 







\onecolumn
\begin{table}[h]
\centering
\begin{tabular}{|l c c|} 
    \hline
    Name & input length (chunk) & embedding dimension  \\ 
    \hline
    embedding reproduction (rep)\cite{yang2018learned}       & 64    & 64 per chunk   \\
    dnabert \cite{ji2021dnabert}                     & 512     & 768 per chunk \\
    prose   \cite{bepler2021learning}                   & 512   & 100 per amino acid   \\
    alphafold  \cite{jumper2021highly}                 & 1024   & 384 per ammino acid\\
    evolutionary scale modeling (esm2) \cite{lin2022language}   & 1024    & 1280 per ammino acid \\  
    \hline
\end{tabular}
\caption{Embedders used in the experiments, their maximum imput length and the dimension of the embedding produced.}
\label{table:data}
\end{table}
\twocolumn

\section{Results}

\onecolumn
\newpage
\FloatBarrier
\bibliographystyle{plain}
\bibliography{bibliography}


\end{document}